File: app/src/main/java/com/arcao/geocaching4locus/base/fragment/AbstractPreferenceFragment.java
package com.arcao.geocaching4locus.base.fragment;

import android.content.SharedPreferences;
import android.os.Bundle;
import android.preference.Preference;
import android.preference.PreferenceFragment;
import android.preference.PreferenceManager;

import com.arcao.geocaching4locus.R;
import com.arcao.geocaching4locus.base.constants.AppConstants;
import com.arcao.geocaching4locus.base.constants.PrefConstants;
import com.arcao.geocaching4locus.base.util.HtmlUtil;

import org.apache.commons.lang3.StringUtils;
import org.oshkimaadziig.george.androidutils.SpanFormatter;

public abstract class AbstractPreferenceFragment extends PreferenceFragment implements SharedPreferences.OnSharedPreferenceChangeListener, PrefConstants {
    public static final String VALUE_HTML_FORMAT = "<font color=\"#FF8000\"><b>%s</b></font>";
    protected SharedPreferences preferences;

    @SuppressWarnings({"unchecked", "unused"})
    protected <P extends Preference> P findPreference(CharSequence key, Class<P> clazz) {
        return (P) super.findPreference(key);
    }

    @Override
    public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String key) {
        // empty
    }

    @Override
    public void onCreate(Bundle paramBundle) {
        super.onCreate(paramBundle);
        preferences = PreferenceManager.getDefaultSharedPreferences(getActivity());
    }

    @Override
    public void onResume() {
        super.onResume();
        getPreferenceScreen().getSharedPreferences().registerOnSharedPreferenceChangeListener(this);
        preparePreference();
    }

    @Override
    public void onPause() {
        super.onPause();
        getPreferenceScreen().getSharedPreferences().unregisterOnSharedPreferenceChangeListener(this);
    }

    protected void preparePreference() {
        if (!getResources().getBoolean(R.bool.preferences_prefer_dual_pane))
            getActivity().setTitle(getPreferenceScreen().getTitle());
    }

    protected CharSequence preparePreferenceSummary(CharSequence value, int resId) {
        CharSequence summary = null;
        if (resId != 0)
            summary = getText(resId);

        if (value != null && value.length() > 0)
            return SpanFormatter.format("%s %s", stylizedValue(value), StringUtils.defaultIfEmpty(summary, ""));
        return StringUtils.defaultIfEmpty(summary, "");
    }

    protected CharSequence stylizedValue(CharSequence value) {
        return SpanFormatter.format(HtmlUtil.fromHtml(VALUE_HTML_FORMAT), value);
    }

    protected void applyPremiumTitleSign(Preference preference) {
        preference.setTitle(String.format("%s %s", preference.getTitle(), AppConstants.PREMIUM_CHARACTER));
    }
}

File: app/src/main/java/com/arcao/geocaching4locus/base/fragment/AbstractPreferenceFragment.kt
package com.arcao.geocaching4locus.base.fragment

import android.content.SharedPreferences
import android.os.Bundle
import androidx.annotation.XmlRes
import androidx.preference.Preference
import androidx.preference.PreferenceFragmentCompat
import androidx.preference.PreferenceManager
import com.arcao.geocaching4locus.base.constants.AppConstants
import com.arcao.geocaching4locus.base.util.HtmlUtil
import org.oshkimaadziig.george.androidutils.SpanFormatter

abstract class AbstractPreferenceFragment : PreferenceFragmentCompat(),
    SharedPreferences.OnSharedPreferenceChangeListener {
    protected lateinit var preferences: SharedPreferences

    @get:XmlRes
    protected abstract val preferenceResource: Int

    protected inline fun <reified P : Preference> preference(key: CharSequence): P {
        return findPreference(key) as P
    }

    override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
        setPreferencesFromResource(preferenceResource, rootKey)
    }

    override fun onSharedPreferenceChanged(sharedPreferences: SharedPreferences, key: String) {
        // empty
    }

    override fun onCreate(paramBundle: Bundle?) {
        super.onCreate(paramBundle)
        preferences = PreferenceManager.getDefaultSharedPreferences(activity)
    }

    override fun onResume() {
        super.onResume()
        preferenceScreen.sharedPreferences.registerOnSharedPreferenceChangeListener(this)
        preparePreference()
    }

    override fun onPause() {
        super.onPause()
        preferenceScreen.sharedPreferences.unregisterOnSharedPreferenceChangeListener(this)
    }

    protected open fun preparePreference() {
        requireActivity().title = preferenceScreen.title
    }

    protected fun preparePreferenceSummary(value: CharSequence?, resId: Int): CharSequence {
        var summary: CharSequence? = null
        if (resId != 0)
            summary = getText(resId)

        return if (!value.isNullOrEmpty()) {
            SpanFormatter.format("%s %s", stylizedValue(value), summary ?: "")
        } else {
            summary ?: ""
        }
    }

    protected fun stylizedValue(value: CharSequence): CharSequence {
        return SpanFormatter.format(HtmlUtil.fromHtml(VALUE_HTML_FORMAT), value)
    }

    protected fun applyPremiumTitleSign(preference: Preference) {
        preference.title = String.format("%s %s", preference.title, AppConstants.PREMIUM_CHARACTER)
    }

    companion object {
        private const val VALUE_HTML_FORMAT = "<font color=\"#FF8000\"><b>%s</b></font>"
    }
}
